extensions/http-curl/sitetosite/HTTPProtocol.cpp

```diff
@@ -0,0 +1,312 @@
+  auto client = create_http_client(uri.str(), "POST");
+
+  client->appendHeader(PROTOCOL_VERSION_HEADER, "1");
+
+  client->setConnectionTimeout(5);
+
+  client->setContentType("application/json");
+  client->appendHeader("Accept: application/json");
+  client->setUseChunkedEncoding();
+  client->setPostFields("");
+  client->submit();
+  peer_->setStream(nullptr);
+  if (client->getResponseCode() == 201) {
+    // parse the headers
+    auto headers = client->getParsedHeaders();
+    auto intent_name = headers["x-location-uri-intent"];
+    if (intent_name == "transaction-url") {
+      auto url = headers["Location"];
+
+      if (IsNullOrEmpty(&url)) {
+        logger_->log_debug("Location is empty");
+      } else {
+
+        org::apache::nifi::minifi::io::CRCStream<SiteToSitePeer> crcstream(peer_.get());
+        auto transaction = std::make_shared<HttpTransaction>(direction, crcstream);
+        transaction->initialize(this, url);
+        auto transactionId = parseTransactionId(url);
+        if (IsNullOrEmpty(transactionId))
+          return nullptr;
+        transaction->setTransactionId(transactionId);
+        std::shared_ptr<minifi::utils::HTTPClient> client;
+        if (transaction->getDirection() == SEND) {
+          client = openConnectionForSending(transaction);
+        } else {
+          client = openConnectionForReceive(transaction);
+          transaction->setDataAvailable(true);
+          // a 201 tells us that data is available. A 200 would mean that nothing is available.
+        }
+
+        client->appendHeader(PROTOCOL_VERSION_HEADER, "1");
+        peer_->setStream(std::unique_ptr<io::DataStream>(new io::HttpStream(client)));
+        transactionID = transaction->getUUIDStr();
+        logger_->log_debug("Created transaction id -%s-", transactionID);
+        known_transactions_[transaction->getUUIDStr()] = transaction;
+        return transaction;
+      }
+    } else {
+      logger_->log_debug("Could not create transaction, intent is %s", intent_name);
+    }
+  } else {
+    logger_->log_debug("Could not create transaction, received %d", client->getResponseCode());
+  }
+  return nullptr;
+}
+
+int HttpSiteToSiteClient::readRespond(const std::shared_ptr<Transaction> &transaction, RespondCode &code, std::string &message) {
+  if (current_code == FINISH_TRANSACTION) {
+
+    if (transaction->getDirection() == SEND) {
+      auto stream = dynamic_cast<io::HttpStream*>(peer_->getStream());
+      stream->closeStream();
+      auto client = stream->getClient();
+      logger_->log_info("Faking it till I makin it 2, response code %d", client->getResponseCode());
+      if (client->getResponseCode() == 202) {
+        code = CONFIRM_TRANSACTION;
+        message = std::string(client->getResponseBody().data(), client->getResponseBody().size());
+      } else {
+        code = UNRECOGNIZED_RESPONSE_CODE;
+      }
+      return 1;
```

Comment:
return -1 for failure

Reply:
If we return -1 we won't have an opportunity to provide an error message. -1 would be for the case where there is a comms error. here we have no need to read and write for respond. But you did remind me to change that name to readResponse.